### ISSUE TITLE
ALC crashes on certain method overloads

### ISSUE THREAD

**Issue by hemisphera** [Microsoft/AL#4035]
I am using ALC 2.0.43900 (taken from the BC OnPrem DVD). The same thing happens when using one taken from a relatively recent docker image (ALC 2.0.46000).

This is a codeunit that I have declared:

````
codeunit 51001 MyCodeunit
{
    procedure CheckDocLine(SalesLine: Record "Sales Line"; WithError: Boolean): Boolean
    begin
    end;
    procedure CheckDocLine(PurchLine: Record "Purchase Line"; WithError: Boolean): Boolean
    begin
    end;
    procedure CheckDocLine(Value1: Date; Value2: Date): Boolean
    begin
    end;
}
````

When I try to compile the extension, ALC throws the following error:

````
error AL0999: Internal error: System.AggregateException: One or more errors occurred. ---> System.InvalidOperationException: Unexpected value 'Boolean' of type 'Microsoft.Dynamics.Nav.CodeAnalysis.NavTypeKind'
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.MethodSymbol.GetParameterTypeHashCode(TypeSymbol parameterType) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\MethodSymbol.cs:line 252
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.MethodSymbol.CalculateMethodId(Boolean isOverloadedMethod) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\MethodSymbol.cs:line 147
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.SourceMethodSymbol.get_Id() in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Source\SourceMethodSymbol.cs:line 65
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.ObjectMembers`1.CheckMembersForDuplicateIds[TMember](ImmutableArray`1 members, DiagnosticBag diagnostics, Func`2 filter, ErrorCode errorCode) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Members\ObjectMembers.cs:line 351
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.ObjectMembers`1.ForceMembersCompleteAndCheckUniqueness(SourceLocation locationOpt, DiagnosticBag diagnostics, CancellationToken cancellationToken, Predicate`1 excludeFromNameUniqueness) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Members\ObjectMembers.cs:line 251
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.ObjectMembers`1.ForceCompleteMembers(SourceLocation locationOpt, CancellationToken cancellationToken, Predicate`1 excludeFromNameUniqueness) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Members\ObjectMembers.cs:line 709
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.ObjectMembers`1.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken, Predicate`1 excludeFromNameUniqueness) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Members\ObjectMembers.cs:line 675
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.SourceCodeunitTypeSymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Source\SourceCodeunitTypeSymbol.cs:line 77
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.SourceModuleSymbol.ForceCompleteMemberByLocation(SourceLocation locationOpt, Symbol member, CancellationToken cancellationToken) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Source\SourceModuleSymbol.cs:line 195
   at Microsoft.Dynamics.Nav.CodeAnalysis.Utilities.UICultureUtilities.<>c__DisplayClass6_0`1.<WithCurrentUICulture>b__0(T param) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Utilities\UICultureUtilities.cs:line 182
   at System.Threading.Tasks.Parallel.<>c__DisplayClass17_0`1.<ForWorker>b__1()
   at System.Threading.Tasks.Task.InnerInvokeWithArg(Task childTask)
   at System.Threading.Tasks.Task.<>c__DisplayClass176_0.<ExecuteSelfReplicating>b__0(Object )
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.For(Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body)
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.SourceModuleSymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Source\SourceModuleSymbol.cs:line 101
   at Microsoft.Dynamics.Nav.CodeAnalysis.Compilation.GetSourceDeclarationDiagnostics(SyntaxTree syntaxTree, Nullable`1 filterSpanWithinTree, Func`4 locationFilterOpt, CancellationToken cancellationToken) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compilation\Compilation.cs:line 930
   at Microsoft.Dynamics.Nav.CodeAnalysis.Compilation.GetDiagnostics(CompilationStage stage, Boolean includeEarlierStages, CancellationToken cancellationToken) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compilation\Compilation.cs:line 616
---> (Inner Exception #0) System.InvalidOperationException: Unexpected value 'Boolean' of type 'Microsoft.Dynamics.Nav.CodeAnalysis.NavTypeKind'
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.MethodSymbol.GetParameterTypeHashCode(TypeSymbol parameterType) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\MethodSymbol.cs:line 252
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.MethodSymbol.CalculateMethodId(Boolean isOverloadedMethod) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\MethodSymbol.cs:line 147
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.SourceMethodSymbol.get_Id() in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Source\SourceMethodSymbol.cs:line 65
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.ObjectMembers`1.CheckMembersForDuplicateIds[TMember](ImmutableArray`1 members, DiagnosticBag diagnostics, Func`2 filter, ErrorCode errorCode) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Members\ObjectMembers.cs:line 351
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.ObjectMembers`1.ForceMembersCompleteAndCheckUniqueness(SourceLocation locationOpt, DiagnosticBag diagnostics, CancellationToken cancellationToken, Predicate`1 excludeFromNameUniqueness) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Members\ObjectMembers.cs:line 251
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.ObjectMembers`1.ForceCompleteMembers(SourceLocation locationOpt, CancellationToken cancellationToken, Predicate`1 excludeFromNameUniqueness) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Members\ObjectMembers.cs:line 709
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.ObjectMembers`1.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken, Predicate`1 excludeFromNameUniqueness) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Members\ObjectMembers.cs:line 675
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.SourceCodeunitTypeSymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Source\SourceCodeunitTypeSymbol.cs:line 77
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.SourceModuleSymbol.ForceCompleteMemberByLocation(SourceLocation locationOpt, Symbol member, CancellationToken cancellationToken) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Source\SourceModuleSymbol.cs:line 195
   at Microsoft.Dynamics.Nav.CodeAnalysis.Utilities.UICultureUtilities.<>c__DisplayClass6_0`1.<WithCurrentUICulture>b__0(T param) in C:\a\669\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Utilities\UICultureUtilities.cs:line 182
   at System.Threading.Tasks.Parallel.<>c__DisplayClass17_0`1.<ForWorker>b__1()
   at System.Threading.Tasks.Task.InnerInvokeWithArg(Task childTask)
   at System.Threading.Tasks.Task.<>c__DisplayClass176_0.<ExecuteSelfReplicating>b__0(Object )<---
````

- When I remove one of the two overloads that take a record (SalesLine or PurchaseLine), then the codeunit compiles just fine. 
- When I add a new overload that takes 1 single integer, the codeunit compiles just fine.

It looks like I cannot have more than one overload where the only thing that changes is the record I pass.

**Comment by hemisphera** [comment]
Addendum:

I have another codeunit where I have the following tow methods:
`procedure CheckDocument(SalesHeader: Record "Sales Header");`
`procedure CheckDocument(PurchHeader: Record "Purchase Header");`

They do compile just fine also.

As soon as I change the signature into
`procedure CheckDocument(SalesHeader: Record "Sales Header"; withError: Boolean);`
`procedure CheckDocument(PurchHeader: Record "Purchase Header; withError: Boolean);`

ALC errors out with the same error as above.

**Comment by hemisphera** [comment]
> 
> 
> May be it would be an option to implement a variant facade?
> https://community.dynamics.com/nav/w/designpatterns/246.variant-facade

It could. But it really **really** should not. I have used variant facades lots and lots of times in C/AL. But that was just because C/AL does not allow function overloads and it was pretty much our only choice to "simulate" that behaviour.

Now that we have overloading in AL, we really shouldn't use this pattern anymore. It has one major downside: we can pass whatever we want into a variant. And therefore we rely on the developer of the fucntion, to actually check what's inside the variant-surprise-box. Too often people forget to do so.

Through overloading we get the same advantages we had with variants. But the difference is: we implicitly limit the possible values we can receive, and checking the variant-surprise-box is no longer up to the developer, but is already handled for us.

Anyway: it's not the point of my issue to discuss variant vs. overload viability. The point of my issue is: There's a feature. It does not work. It should be fixed.

**Comment by waldo1001** [comment]
Just trying to help :-).  
I agree it's a bug.

**Comment by BartPermentier** [comment]
This bug is not fixed in AL Language 2.1.109895!
Compiler error when trying to build:
[Error - 11:15:19 AM] Please report this issue to https://github.com/microsoft/al/issues including information on how to reproduce it, if possible.
Processing of message 'al/createPackage' failed with error: 'Unexpected value 'Boolean' of type 'Microsoft.Dynamics.Nav.CodeAnalysis.NavTypeKind''
Details:
System.InvalidOperationException: Unexpected value 'Boolean' of type 'Microsoft.Dynamics.Nav.CodeAnalysis.NavTypeKind'
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.MethodSymbol.GetParameterTypeHashCode(TypeSymbol parameterType) in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\MethodSymbol.cs:line 252
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.MethodSymbol.CalculateMethodId(Boolean isOverloadedMethod) in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\MethodSymbol.cs:line 147
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.SourceMethodSymbol.get_Id() in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Source\SourceMethodSymbol.cs:line 65
   at Microsoft.Dynamics.Nav.CodeAnalysis.SymbolExtensions.GetOverloadedMethodMetadataNameSuffix(MethodSymbol method) in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\KindsAndExtensions\SymbolExtensions.cs:line 180
   at Microsoft.Dynamics.Nav.CodeAnalysis.Symbols.SourceMethodSymbol.get_MetadataName() in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Symbols\Source\SourceMethodSymbol.cs:line 43
   at Microsoft.Dynamics.Nav.CodeAnalysis.Emit.MethodCodeGenerator.GetMethodScopeName(MethodSymbol methodSymbol) in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Emitter\CodeGen\MethodCodeGenerator.cs:line 105
   at Microsoft.Dynamics.Nav.CodeAnalysis.Emit.MethodCodeGenerator..ctor(MethodSymbol method, BoundBlock boundBlock, ModuleBuilder moduleBuilder, CompilationOptions compilationOptions) in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Emitter\CodeGen\MethodCodeGenerator.cs:line 78
   at Microsoft.Dynamics.Nav.CodeAnalysis.MethodCompiler.CompileMethod(MethodSymbol methodSymbol) in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compiler\MethodCompiler.cs:line 455
   at Microsoft.Dynamics.Nav.CodeAnalysis.MethodCompiler.CompileMembers(ContainerSymbol containingType) in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compiler\MethodCompiler.cs:line 328
   at Microsoft.Dynamics.Nav.CodeAnalysis.MethodCompiler.CompileObject(ObjectTypeSymbol containingType) in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compiler\MethodCompiler.cs:line 218
   at Microsoft.Dynamics.Nav.CodeAnalysis.MethodCompiler.<>c__DisplayClass14_0.<CompileObjectAsTask>b__0() in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compiler\MethodCompiler.cs:line 202
   at Microsoft.Dynamics.Nav.CodeAnalysis.Utilities.UICultureUtilities.<>c__DisplayClass5_0.<WithCurrentUICulture>b__0() in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Utilities\UICultureUtilities.cs:line 152
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Dynamics.Nav.CodeAnalysis.MethodCompiler.WaitForWorkers() in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compiler\MethodCompiler.cs:line 162
   at Microsoft.Dynamics.Nav.CodeAnalysis.MethodCompiler.CompileMethodBodies(Compilation compilation, ModuleBuilder moduleBeingBuiltOpt, Boolean hasDeclarationErrors, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compiler\MethodCompiler.cs:line 146
   at Microsoft.Dynamics.Nav.CodeAnalysis.Compilation.CompileImpl(ModuleBuilder moduleBuilder, Boolean emittingPdb, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compilation\Compilation.cs:line 886
   at Microsoft.Dynamics.Nav.CodeAnalysis.Compilation.Compile(ModuleBuilder moduleBuilder, Boolean emittingPdb, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compilation\Compilation.cs:line 861
   at Microsoft.Dynamics.Nav.CodeAnalysis.CommandLine.CommonCompiler.PackageTheModule(Compilation compilation, CompilerTextWriter consoleOutput, CancellationToken cancellationToken, ErrorLogger errorLogger, AnalyzerDriver analyzerDriver) in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\CommandLine\CommonCompiler.cs:line 577
   at Microsoft.Dynamics.Nav.CodeAnalysis.CommandLine.CommonCompiler.RunCore(CompilerTextWriter consoleOutput, ErrorLogger errorLogger, CancellationToken cancellationToken) in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\CommandLine\CommonCompiler.cs:line 435
   at Microsoft.Dynamics.Nav.CodeAnalysis.CommandLine.CommonCompiler.Run(CompilerTextWriter consoleOutput, CancellationToken cancellationToken) in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\CommandLine\CommonCompiler.cs:line 313
   at Microsoft.Dynamics.Nav.EditorServices.Protocol.LanguageServer.Extensions.CreatePackageRequestHandler.<HandleAsync>d__4.MoveNext() in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.EditorServices.Protocol\LanguageServer\Extensions\CreatePackageRequestHandler.cs:line 47
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Dynamics.Nav.EditorServices.Protocol.MessageProtocol.RequestHandlerBase`1.<HandleAsync>d__10.MoveNext() in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.EditorServices.Protocol\MessageProtocol\RequestHandlerBase.cs:line 45
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Dynamics.Nav.EditorServices.Protocol.RequestRegistry.<Process>d__6.MoveNext() in F:\a\35\s\source\Prod\Microsoft.Dynamics.Nav.EditorServices.Protocol\Endpoints\RequestRegistry.cs:line 59
![image](https://user-images.githubusercontent.com/24684493/54816463-8f9fdc00-4c94-11e9-94c8-d7a366269ddc.png)
It builds just fine if I remove the ModifyQty Boolean.
